pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -0,0 +1,382 @@
+            raise ValueError(
+                'This manager has been closed and can not be re-used.')
+
+        self._callback = functools.partial(_wrap_callback_errors, callback)
+
+        # Start the thread to pass the requests.
+        self._dispatcher = dispatcher.Dispatcher(self, self._scheduler.queue)
+        self._dispatcher.start()
+
+        # Start consuming messages.
+        self._rpc = bidi.ResumableBidiRpc(
+            start_rpc=self._client.api.streaming_pull,
+            initial_request=self._get_initial_request,
+            should_recover=self._should_recover)
+        self._rpc.add_done_callback(self._on_rpc_done)
+        self._consumer = bidi.BackgroundConsumer(
+            self._rpc, self._on_response)
+        self._consumer.start()
+
+        # Start the lease maintainer thread.
+        self._leaser = leaser.Leaser(self)
+        self._leaser.start()
+
+    def close(self, reason=None):
+        """Stop consuming messages and shutdown all helper threads.
+
+        This method is idempotent. Additional calls will have no effect.
+
+        Args:
+            reason (Any): The reason to close this. If None, this is considered
+                an "intentional" shutdown. This is passed to the callbacks
+                specified via :meth:`add_close_callback`.
+        """
+        with self._closing:
+            if self._closed:
+                return
+
+            # Stop consuming messages.
+            if self.is_active:
+                _LOGGER.debug('Stopping consumer.')
+                self._consumer.stop()
+            self._consumer = None
+
+            # Shutdown all helper threads
+            _LOGGER.debug('Stopping scheduler.')
+            self._scheduler.shutdown()
+            self._scheduler = None
+            _LOGGER.debug('Stopping leaser.')
+            self._leaser.stop()
+            self._leaser = None
+            _LOGGER.debug('Stopping dispatcher.')
+            self._dispatcher.stop()
+            self._dispatcher = None
+
+            self._rpc = None
+            self._closed = True
+            _LOGGER.debug('Finished stopping manager.')
+
+            for callback in self._close_callbacks:
+                callback(self, reason)
+
+    def _get_initial_request(self):
+        """Return the initial request for the RPC.
+
+        This defines the initial request that must always be sent to Pub/Sub
+        immediately upon opening the subscription.
+
+        Returns:
+            google.cloud.pubsub_v1.types.StreamingPullRequest: A request
+            suitable for being the first request on the stream (and not
+            suitable for any other purpose).
+        """
+        # Any ack IDs that are under lease management need to have their
+        # deadline extended immediately.
+        lease_ids = self._leaser.ack_ids
+
+        # Put the request together.
+        request = types.StreamingPullRequest(
+            modify_deadline_ack_ids=list(lease_ids),
```

Comment:
If not required to create as list, should remove.

Reply:
We have to materialize the list at some point for protobuf, might as well be explicit here.